sql-plugin/src/main/scala/com/nvidia/spark/rapids/spill/SpillFramework.scala

```diff
@@ -0,0 +1,1606 @@
+ * intended store (`DeviceSpillableHandle` incRefCounts an object if it is in the device store),
+ * and otherwise it will create a new copy from the spilled version and hand it to the user.
+ * Any time a user calls `materialize`, they are responsible for closing the returned object.
+ *
+ * Spilling:
+ *
+ * A `SpillableHandle` will track an object in a specific store (`DeviceSpillable` tracks
+ * device "intended" objects) for example. If the handle is asked to spill, it is the handle's
+ * responsibility to initiate that spill, and to track the spilled handle (a device spillable
+ * would have a `host` handle, which tracks the host spilled object).
+ *
+ * A cascade of spills can occur device -> host -> disk, given that host allocations can fail, or
+ * could not fit in the SpillableHostStore's limit (if defined). In this case, the call to spill
+ * will either create a host handle tracking an object on the host store (if we made room), or it
+ * will create a host handle that points to a disk handle, tracking a file on disk.
+ *
+ * If the disk is full, we do not handle this in any special way. We expect this to be a
+ * terminal state to the executor. Every handle spills to its own file on disk, identified
+ * as a "temporary block" `BlockId` from Spark.
+ *
+ * Notes on locking:
+ *
+ * All stores use a concurrent hash map to store instances of `StoreHandle`. The only store
+ * with extra locking is the `SpillableHostStore`, to maintain a `totalSize` number that is
+ * used to figure out cheaply when it is full.
+ *
+ * Handles hold a lock to protect the user against when it is either in the middle of
+ * spilling, or closed. Most handles, except for disk handles, hold a reference to an object
+ * in their respective store (`SpillableDeviceBufferHandle` has a `dev` reference that holds
+ * a `DeviceMemoryBuffer`, and a `host` reference to `SpillableHostBufferHandle` that is only
+ * set if spilled), and the handle guarantees that one of these will be set, or none if the handle
+ * is closed.
+ *
+ * We hold the handle lock when we are spilling (performing IO). That means that no other consumer
+ * can access this spillable device handle while it is being spilled, including a second thread
+ * trying to spill and asking each of the handles whether they are spillable or not, as that
+ * requires the handle lock. We will relax this likely in follow on work.
+ *
+ * We never hold a store-wide coarse grain lock in the stores when we do IO.
+ */
+
+/**
+ * Common interface for all handles in the spill framework.
+ */
+trait StoreHandle extends AutoCloseable {
+  /**
+   * Approximate size of this handle, used in two scenarios:
+   * - Used from the host store to figure out how much host memory total it is tracking.
+   * - If sizeInBytes is 0, the object is tracked by the stores so it can be
+   *   removed on shutdown, or by handle.close, but 0-byte handles are not spillable.
+   */
+  val sizeInBytes: Long
```

Comment:
Also this is a val, but if something spills wouldn't it change to 0 from whatever it was before? Shouldn't it be a def?

Reply:
So it is, and it isn't. That's the issue with `sizeInBytes`. I am going to rework that and make it clearer. I think something like `approxSizeInBytes` makes sense, or `trackedSizeInBytes`, and it should be `private[spill]` going with the other things that are marked that way. But some handles do have a size that is not approximate, and IS used for creating buffers, so I kind of want that size reported with a different API, specific to each handle. 

It is a val because I didn't see a point in changing it. If the object is spilled that is determined by `dev`, or `host` being empty.

Reply:
It is also a val, because I wanted to set `spillable` at construction for 0-byte handles. Any other handle is spillable by definition at the level of the trait `SpillableHandle`, but if it's not a val then there's a chance that `approxSizeInBytes` is not ready at construction time, so I don't know if I can call it reliably.

Reply:
agree this is confusing. The val part is because I want to mark objects that are 0 sized as not spillable, and I'd like to do that at construction time. It removes some locking that otherwise I have to do.

I have renamed this `approxSizeInBytes` and left `sizeInBytes` public in handles that do support this non-approximate value. I default this so in most cases `approxSizeInBytes = sizeInBytes`, but I do think this makes it clearer.